fhir-server/src/test/java/au/csiro/pathling/fhirpath/function/subsumes/ClosureTest.java

```diff
@@ -0,0 +1,148 @@
+/*
+ * Copyright © 2018-2020, Commonwealth Scientific and Industrial Research
+ * Organisation (CSIRO) ABN 41 687 119 230. Licensed under the CSIRO Open Source
+ * Software Licence Agreement.
+ */
+
+package au.csiro.pathling.fhirpath.function.subsumes;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+import au.csiro.pathling.fhirpath.encoding.SimpleCoding;
+import au.csiro.pathling.fhirpath.function.subsumes.Closure.CodingSet;
+import java.util.*;
+import org.junit.Test;
+import org.junit.jupiter.api.Tag;
+
+@Tag("UnitTest")
+public class ClosureTest {
+
+  public static final SimpleCoding CODING1_UNVERSIONED = new SimpleCoding("uuid:system1", "code");
+  public static final SimpleCoding CODING1_VERSION1 =
+      new SimpleCoding("uuid:system1", "code", "version1");
+  public static final SimpleCoding CODING1_VERSION2 =
+      new SimpleCoding("uuid:system1", "code", "version2");
+
+  public static final SimpleCoding CODING2_UNVERSIONED = new SimpleCoding("uuid:system2", "code");
+  public static final SimpleCoding CODING2_VERSION1 =
+      new SimpleCoding("uuid:system2", "code", "version1");
+  public static final SimpleCoding CODING2_VERSION2 =
+      new SimpleCoding("uuid:system2", "code", "version2");
+
+  public static final SimpleCoding CODING3_UNVERSIONED = new SimpleCoding("uuid:system1", "code1");
+  public static final SimpleCoding CODING3_VERSION1 =
+      new SimpleCoding("uuid:system1", "code1", "version1");
+  public static final SimpleCoding CODING3_VERSION2 =
+      new SimpleCoding("uuid:system1", "code1", "version2");
+
+  private static Set<SimpleCoding> setOf(SimpleCoding... codings) {
+    return new HashSet<SimpleCoding>(Arrays.asList(codings));
+  }
+
+  @Test
+  public void testVersionedCodingSet() {
+    CodingSet versionedCodingSet = new Closure.CodingSet(setOf(CODING1_VERSION1));
+    assertThat(versionedCodingSet.contains(CODING1_UNVERSIONED)).isTrue();
+    assertThat(versionedCodingSet.contains(CODING1_VERSION1)).isTrue();
+    assertThat(versionedCodingSet.contains(CODING1_VERSION2)).isFalse();
+    assertThat(versionedCodingSet.contains(CODING2_VERSION1)).isFalse();
+  }
```

Comment:
I changed the semantics for Coding equality as part of this refactor. You can see what I landed on in the implementation of the `CodingsEqual` UDF.

Now the version agnostic test is applied if the version is missing from either Coding - so it's a bit more lenient.

I think the reason I changed this is because the terminology server sometimes adds in a version to a code, based on the version of the code system it has loaded. This causes problems with matching the input Codings from the data with responses from the terminology service.

The other reason is that, on reflection, I think that the previous equality test was probably too strict for most real-world query use cases - but I guess we will gather more feedback about this from users as time goes on.

Reply:
I think that the implementation of CodingSet.contatins follows this lenient semantics and the expectations do reflect it as well, e.g. if the actual element is versioned returns true for both unversioned coding and the coding with matching version:

```
  public void testVersionedCodingSet() {
    CodingSet versionedCodingSet = new Closure.CodingSet(setOf(CODING1_VERSION1));
    assertTrue(versionedCodingSet.contains(CODING1_UNVERSIONED));
    assertTrue(versionedCodingSet.contains(CODING1_VERSION1));
```

If the actual coding in the set is unversioned the contains returns true for an unversioned coding or matching coding with any version:
```
  @Test
  public void testUncersionedCodingSet() {
    CodingSet unversionedCodingSet = new Closure.CodingSet(setOf(CODING1_UNVERSIONED));
    assertTrue(unversionedCodingSet.contains(CODING1_UNVERSIONED));
    assertTrue(unversionedCodingSet.contains(CODING1_VERSION1));
    assertTrue(unversionedCodingSet.contains(CODING1_VERSION2));
```

Reply:
I've just gone through all the expectations and they look correct, thanks.